File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/youtube/service/AtomParser.java
package com.github.bumblebee.command.youtube.service;

import com.github.bumblebee.command.youtube.entity.AtomFeed;
import org.springframework.stereotype.Component;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import javax.xml.transform.stream.StreamSource;
import java.io.StringReader;

@Component
public class AtomParser {

    public AtomFeed parse(String input) throws JAXBException {
        JAXBContext context = JAXBContext.newInstance(AtomFeed.class);
        Unmarshaller unmarshaller = context.createUnmarshaller();
        JAXBElement<AtomFeed> root = unmarshaller.unmarshal(new StreamSource(new StringReader(input)), AtomFeed.class);
        return root.getValue();
    }
}

File: telegram-bot-bumblebee-core/src/main/kotlin/com/github/bumblebee/command/youtube/service/AtomParser.kt
package com.github.bumblebee.command.youtube.service

import com.github.bumblebee.command.youtube.entity.VideoNotification
import org.springframework.stereotype.Component
import org.w3c.dom.Element
import org.xml.sax.InputSource
import java.io.ByteArrayInputStream
import javax.xml.parsers.DocumentBuilderFactory

@Component
class AtomParser {

    fun parse(xml: String): VideoNotification? {
        val dbFactory = DocumentBuilderFactory.newInstance()
        val dBuilder = dbFactory.newDocumentBuilder()
        val doc = dBuilder.parse(InputSource(ByteArrayInputStream(xml.toByteArray())))
        doc.documentElement.normalize()

        val entry = doc.documentElement.getElementsByTagName("entry").item(0) as? Element
        val videoId = entry?.getElementsByTagName("yt:videoId")?.item(0)?.textContent
        val channelId = entry?.getElementsByTagName("yt:channelId")?.item(0)?.textContent

        return if (videoId != null && channelId != null)
            VideoNotification(videoId, channelId)
        else null
    }
}
